forge/src/debugger.rs

```diff
@@ -0,0 +1,212 @@
+    /// The program counter
+    pub pc: usize,
+    /// Stack *prior* to running the associated opcode
+    pub stack: Vec<U256>,
+    /// Memory *prior* to running the associated opcode
+    pub memory: Memory,
+    /// Opcode to be executed
+    pub instruction: Instruction,
+    /// Optional bytes that are being pushed onto the stack
+    pub push_bytes: Option<Vec<u8>>,
+    /// Instruction counter, used to map this instruction to source code
+    pub ic: usize,
+    /// Cumulative gas usage
+    pub total_gas_used: u64,
+}
+
+impl Default for DebugStep {
+    fn default() -> Self {
+        Self {
+            pc: 0,
+            stack: vec![],
+            memory: Memory::new(),
+            instruction: Instruction::OpCode(revm::opcode::INVALID),
+            push_bytes: None,
+            ic: 0,
+            total_gas_used: 0,
+        }
+    }
+}
+
+impl DebugStep {
+    /// Pretty print the step's opcode
+    pub fn pretty_opcode(&self) -> String {
+        if let Some(push_bytes) = &self.push_bytes {
+            format!("{}(0x{})", self.instruction, hex::encode(push_bytes))
+        } else {
+            self.instruction.to_string()
+        }
+    }
+}
+
+impl Display for DebugStep {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        if let Some(push_bytes) = &self.push_bytes {
+            write!(
+                f,
+                "pc: {:?}\nop: {}(0x{})\nstack: {:#?}\nmemory: 0x{}\n\n",
+                self.pc,
+                self.instruction,
+                hex::encode(push_bytes),
+                self.stack,
+                hex::encode(self.memory.data())
+            )
+        } else {
+            write!(
+                f,
+                "pc: {:?}\nop: {}\nstack: {:#?}\nmemory: 0x{}\n\n",
+                self.pc,
+                self.instruction,
+                self.stack,
+                hex::encode(self.memory.data())
+            )
+        }
+    }
+}
+
+#[derive(Debug, Copy, Clone)]
+pub enum Instruction {
+    OpCode(u8),
+    Cheatcode([u8; 4]),
+}
+
+impl From<u8> for Instruction {
+    fn from(op: u8) -> Instruction {
+        Instruction::OpCode(op)
+    }
+}
+
+impl Display for Instruction {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Instruction::OpCode(op) => write!(
+                f,
+                "{}",
+                OpCode::try_from_u8(*op).map_or_else(
+                    || format!("UNDEFINED(0x{:02x})", op),
+                    |opcode| opcode.as_str().to_string(),
+                )
+            ),
+            Instruction::Cheatcode(cheat) => write!(
+                f,
+                "VM_{}",
+                &*HEVM_ABI
+                    .functions()
+                    .find(|func| func.short_signature() == *cheat)
+                    .expect("unknown cheatcode found in debugger")
+                    .name
+                    .to_uppercase()
+            ),
```

Comment:
I got rid of the `CheatOp` struct. Collecting debug info is a flag-gated one time cost and it is fast enough that we can focus on maintanability instead - this just makes it so we don't have to touch the debugger when we add new cheatscodes.